tokio/src/sync/once_cell.rs

```diff
@@ -464,6 +629,131 @@ unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
 // it's safe to send it to another thread
 unsafe impl<T: Send> Send for OnceCell<T> {}
 
+/// An entry in the wait queue.
+struct Waiter {
+    /// If the waiter is a cell initializer.
+    is_initializer: bool,
+    /// Waiting task.
+    waker: Option<Waker>,
+    /// Intrusive linked-list pointers.
+    pointers: linked_list::Pointers<Waiter>,
+    /// Should not be `Unpin`.
+    _p: PhantomPinned,
+}
+
+impl Waiter {
+    fn new(is_initializer: bool) -> Self {
+        Self {
+            is_initializer,
+            waker: None,
+            pointers: linked_list::Pointers::new(),
+            _p: PhantomPinned,
+        }
+    }
+}
+
+// SAFETY: `Waiter` is forced to be !Unpin.
+unsafe impl linked_list::Link for Waiter {
+    type Handle = NonNull<Waiter>;
+    type Target = Waiter;
+
+    fn as_raw(handle: &NonNull<Waiter>) -> NonNull<Waiter> {
+        *handle
+    }
+
+    unsafe fn from_raw(ptr: NonNull<Waiter>) -> NonNull<Waiter> {
+        ptr
+    }
+
+    unsafe fn pointers(target: NonNull<Waiter>) -> NonNull<linked_list::Pointers<Waiter>> {
+        Waiter::addr_of_pointers(target)
+    }
+}
+
+generate_addr_of_methods! {
+    impl<> Waiter {
+        unsafe fn addr_of_pointers(self: NonNull<Self>) -> NonNull<linked_list::Pointers<Waiter>> {
+            &self.pointers
+        }
+    }
+}
+
+/// Wait for the cell to be initialized, or the state to be unlocked
+/// in the case of an initializer.
+struct WaitFuture<'a, T> {
+    node: Waiter,
+    cell: &'a OnceCell<T>,
+    queued: bool,
```

Comment:
This `queued` field is just for the optimization in `Drop` to skip the mutex? I don't think you can actually ever make use of that optimization as currently implemented, so you can just drop the field.

Reply:
Could develop about why the fields would never been used?
If `wait_initialized` is called while the cell is already initialized, `poll_wait` will return `Poll::Ready` and `queued` field will never be set to `true`.
Tracing the early return of `drop` in tests shows that it is executed.

Reply:
I guess you are right.